### PR TITLE
[1pt] PR: Separate hydroconditioning for branch flow direction generation

### PR DESCRIPTION
**Pull request by RyanSpies-NOAA**
These changes address some known issues where the DEM derived flowlines follow the incorrect flow path (address issues with stream order 1 and 2 only). The revised code adds a new workflow to generate a new flow direction raster separately for input to the `run_by_branch.sh` workflow (branch 0 remains unchanged). This modification helps ensure that the DEM derived flowlines follow the desired NWM flow line when generating the DEM derived flowlines at the branch level. Addresses #823 

### Changes  
- `config/deny_branch_zero.lst`: removed `LandSea_subset_{}.tif` and `flowdir_d8_burned_filled_{}.tif` from the "keep" list as these files are now kept in the huc root folder.
- `config/deny_unit.lst`: added file cleanups for newly generated branch input files stored in the huc root folder (`dem_burned.tif`, `dem_burned_filled.tif`, `flowdir_d8_burned_filled.tif`, `flows_grid_boolean.tif`, `wbd_buffered_streams.gpkg`)
- `src/clip_vectors_to_wbd.py`: saving the `wbd_streams_buffer` as an output gpkg for input to `derive_level_paths.py`
- `src/derive_level_paths.py`: added a new step to clip the `out_stream_network_dissolved` with the `buffer_wbd_streams` polygon. this resolves errors with the edge case scenarios where a NWM flow line intersects the WBD buffer polygon
- `src/run_unit_wb.sh`: Introduce new processing steps to generate separate outputs for input to branch 0 vs. all other branches. Remove the branch zero `outputs_cleanup.py` as the branches are no longer pointing to files stored in the branch 0 directory (stored in huc directory)
   - Rasterize reach boolean (1 & 0) for all branches (not branch 0): using the `nwm_subset_streams_levelPaths_dissolved.gpkg` to define the branch levelpath flow lines
   - AGREEDEM reconditioning for all branches (not branch 0)
   - Pit remove burned DEM for all branches (not branch 0)
   - D8 flow direction generation for all branches (not branch 0)
- `src/run_by_branch.sh`: changed `clip_rasters_to_branches.py` input file location for `$tempHucDataDir/flowdir_d8_burned_filled.tif` (newly created file)

### Testing
- Successfully completed an alpha test domain run with no unit or branch errors

image

### Notes
Results from the alpha test
mcc_nws_comp
mcc_usgs_comp
mcc_ble_comp
mcc_ifc_comp

### Screenshots
image

image

image

image

### Checklist

_You may update this checklist before and/or after creating the PR. If you're unsure about any of them, please ask, we're here to help! These items are what we are going to look for before merging your code._

- [x] Informative and human-readable title, using the format: `[_pt] PR: <description>`
- [x] Links are provided if this PR resolves an issue, or depends on another other PR
- [x] If submitting a PR to the `dev` branch (the default branch), you have a descriptive [Feature Branch](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) name using the format: `dev-<description-of-change>` (e.g.: `dev-revise-levee-masking`)
- [x] Changes are limited to a single goal (no scope creep)
- [x] The feature branch you're submitting as a PR is up to date (merged) with the latest `dev` branch
- [x] Changes adhere to [PEP-8 Style Guidelines](https://peps.python.org/pep-0008/)
- [x] Any _change_ in functionality is tested
- [ ] Passes all unit tests locally (inside interactive Docker container, at `/foss_fim/`, run: `pytest unit_tests/`)
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Project documentation has been updated ([CHANGELOG](/docs/CHANGELOG.md) and/or [README](/README.md))
- [x] [Reviewers requested](https://help.github.com/articles/requesting-a-pull-request-review/)

### CODE REVIEW

**Comment by mluck** [comment]
The inundation looks significantly better, but there seems to be a bit of regression in MCC in many of the metrics (NWS, USGS, and IFC). What's causing that?

**Comment by mluck** [comment]
This PR probably addresses most of the issues where level paths diverge onto lower order streams, but only if they are orders 1 or 2 (correct?). However, this won't fix a higher order stream from following a lower order stream if the lower order stream is 3 or above. Do we know how prevalent this case is, or if it happens at all?

**Comment by RyanSpies-NOAA** [comment]
Good point Matt. Correct - this fix would only address incorrect DEM derived flow lines that incorrectly follow NWM stream lines orders 1 & 2. From the ~15 HUCs that I qualitatively inspected, the orders 1 & 2 were the only identifiable culprits. It appears the majority of cases seem to occur where the NWM has a stream order 1 that loops off of the main channel ("abandoned channels"). I'm anticipating that there are likely other culprits (orders > 2), but they would appear to be much less common than the smaller order stream trip ups. The variable burn depth approach based on NWM order, may end up being a more hardened solution for this - that approach may also address the problem for branch 0.

image

**Review by mluck — APPROVED** [review]
The PR does a good job of addressing the error where the DEM-derived stream leaves the level path channel and follows a lower order segment. Prior to the PR, the DEM-derived reaches follow the red dashed line, but afterwards it correctly stays on the dark blue line.
image

The PR also fixes the negative REM problem in HUC 18030007 noted in https://github.com/NOAA-OWP/inundation-mapping/issues/860#issuecomment-1507693096.

Tested on HUCs 08020302 and 18030007.

File: docs/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,24 @@ All notable changes to this project will be documented in this file.
 We follow the [Semantic Versioning 2.0.0](http://semver.org/) format.
 
 
+## v4.3.X.X - 2023-05-12 - [PR#903](https://github.com/NOAA-OWP/inundation-mapping/pull/903)
+
+These changes address some known issues where the DEM derived flowlines follow the incorrect flow path (smaller stream order). The revised code adds a new workflow to generate a new flow direction raster separately for input to the `run_by_branch.sh` workflow (branch 0 remains unchanged). This modification helps ensure that the DEM derived flowlines follow the desired NWM flow line when generating the DEM derived flowlines at the branch level. Addresses #823
```
[the remark this answers is not in the document]
Comment (RyanSpies-NOAA) [reply docs/CHANGELOG.md:7]: added

**Review by RobHanna-NOAA — APPROVED** [review]

**Pull request merged**